inferno-core/src/Inferno/Infer.hs

```diff
@@ -784,23 +784,57 @@ infer expr =
                       [ tyConstr t1 (t2 `TArr` tv) [ExpectedFunction tyCls (t2 `TArr` tv) t1 $ blockPosition e1]
                       ]
                 )
-
+        LetAnnot p1 loc x pT t p2 e1 p3 e2 -> do
+          (e1', ImplType i1 t1, c1) <- infer e1
+          (tcs, (ImplType iT tT)) <- instantiate t
+          let tyCls = Set.fromList $ map snd $ rights $ Set.toList c1
+          attachTypeToPosition
+            (elementPosition loc $ Expl x)
+            TypeMetadata
+              { identExpr = Var () () LocalScope $ Expl x,
+                -- ty = (tyCls, ImplType i1 t1),
+                ty = (tcs, (ImplType iT tT)),
+                docs = Nothing
+              }
+          let newEnv =
+                ( x,
+                  TypeMetadata
+                    { identExpr = Var () () LocalScope $ Expl x,
+                      -- ty = ForallTC [] tyCls $ ImplType i1 t1,
+                      ty = ForallTC [] tcs (ImplType iT tT),
+                      docs = Nothing
+                    }
+                )
+          (e2', ImplType i2 t2, c2) <- inEnv newEnv $ infer e2
+          let (isMerged, ics) = mergeImplicitMaps (blockPosition expr) [i1, i2, iT]
+          return
+            ( LetAnnot p1 loc x pT t p2 e1' p3 e2',
+              ImplType isMerged t2,
+              Set.fromList ics
+                `Set.union` c1
+                `Set.union` c2
+                -- Type of e1 == type annotation
+                `Set.union` Set.fromList [tyConstr t1 tT [UnificationFail tyCls t1 tT $ blockPosition e1]]
+                -- Type class constraints from type annotation TODO filter out reps?
+                `Set.union` Set.map (Right . (exprLoc,)) tcs
```

Comment:
@goodlyrottenapple I saw that the code for `Var` was [filtering out](https://github.com/plow-technologies/inferno/blob/af58127ad12758185d78c313ff911d873a9e90bb/inferno-core/src/Inferno/Infer.hs#L562) the `rep` typeclass constraints, why does it do that and would we need to do something similar here?

Reply:
The rep typeclass doesn't really constrain the type in any way, because we have an an instance for every type, so it was easier to just remove it than pass it to the constraint solver and deal with it there. As far as i remember, the inference modifies a type-checked term and inserts `TyRep` wherever we need the runtime information, so it should be safe to remove the ty-rep typeclass information at call sites of already type-checked terms?

As to the call site here, you shouldn't allow the user to write a tyrep typeclass in the first place, so probably no need to try to filter it out?